compiler/hash-lexer/src/error.rs

```diff
@@ -58,38 +61,63 @@ pub enum LexerErrorKind {
     Unclosed(Delimiter),
 }
 
-// For now, we don't use the reporting capabilities within the lexer and just
-// let the `hash-parser` crate handle the reporting. This is partially due to
-// the fact that the current infrastructure within the compiler will lex and
-// parse at the same time within the thread pool. Ideally, instead of passing
-// `<Lexer|Parser>Error` around, we could just use the `Report` type which is
-// agnostic and can also be sent as messages around the sub-systems.
-
-// impl LexerError {
-//     pub fn create_report(&self, source_id: SourceId) -> Report {
-//         let mut builder = ReportBuilder::new();
-
-//         builder
-//             .with_kind(ReportKind::Error)
-//             .with_message("Failed to parse")
-//             .add_element(ReportElement::CodeBlock(ReportCodeBlock::new(
-//                 SourceLocation {
-//                     span: self.span,
-//                     source_id,
-//                 },
-//                 "here",
-//             )))
-//             .add_element(ReportElement::Note(ReportNote::new(
-//                 ReportNoteKind::Note,
-//                 self.to_string(),
-//             )));
-
-//         builder.build().unwrap()
-//     }
-// }
-
-// impl From<LexerErrorWrapper> for Report {
-//     fn from(LexerErrorWrapper(source_id, err): LexerErrorWrapper) -> Self {
-//         err.create_report(source_id)
-//     }
-// }
+impl From<LexerError> for Report {
+    fn from(err: LexerError) -> Self {
+        let mut builder = ReportBuilder::new();
+
+        builder
+            .with_kind(ReportKind::Error)
+            .with_message(err.to_string())
+            .add_element(ReportElement::CodeBlock(ReportCodeBlock::new(err.location, "here")));
+
+        builder.build()
+    }
+}
+
+/// Lexer error store, the lexer stores an internal store so that
+/// it can implement [Diagnostics::DiagnosticsStore]
+#[derive(Default)]
+pub struct LexerDiagnostics {
+    errors: Vec<LexerError>,
+}
+
+impl Diagnostics<LexerError, LexerError> for Lexer<'_> {
```

Comment:
Maybe use Infallible for the warning ?

Reply:
But the `Diagnostics` trait specifies that `W` needs to implement `Into<Report` which `Infallible` does not

Reply:
Okay then maybe enum LexerWarning {}

Reply:
Ive left it as Infallible for now, im sure `LexerWarning` will be added in the future, but let’s keep it like this for now?